Content.Server/GameObjects/Components/Movement/ServerPortalComponent.cs

```diff
@@ -0,0 +1,207 @@
+            // How long before a specific person can go back into it
+            serializer.DataField(ref _individualPortalCooldown, "individual_cooldown", 3.0f);
+            // How long before anyone can go in it
+            serializer.DataField(ref _overallPortalCooldown, "overall_cooldown", 1.0f);
+            serializer.DataField(ref _departureSound, "departure_sound", "/Audio/effects/teleport_departure.ogg");
+            serializer.DataField(ref _arrivalSound, "arrival_sound", "/Audio/effects/teleport_arrival.ogg");
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _appearanceComponent = Owner.GetComponent<AppearanceComponent>();
+        }
+
+        public override void OnAdd()
+        {
+            // This will blow up an entity it's attached to
+            base.OnAdd();
+            if (Owner.TryGetComponent<CollidableComponent>(out var collide))
+            {
+                collide.IsHardCollidable = false;
+            }
+
+            _state = PortalState.UnableToTeleport;
+            if (_aliveTime > 0)
+            {
+                Timer.Spawn(TimeSpan.FromSeconds(_aliveTime), () => Owner.Delete());
+            }
+        }
+
+        public override void OnRemove()
+        {
+            _appearanceComponent = null;
+
+            base.OnRemove();
+        }
+
+        public bool CanBeConnected()
+        {
+            if (_connectingTeleporter == null)
+            {
+                return true;
+            }
+            return false;
+        }
+
+        public void TryConnectPortal(IEntity otherPortal)
+        {
+            // TODO: Refactor to use out or w/e
+            if (otherPortal.TryGetComponent<ServerPortalComponent>(out var connectedPortal) && connectedPortal.CanBeConnected())
+            {
+                _connectingTeleporter = otherPortal;
+                connectedPortal._connectingTeleporter = Owner;
+                TryChangeState(PortalState.Pending);
+            }
+        }
+
+        public void TryChangeState(PortalState targetState)
+        {
+            if (Owner == null)
+            {
+                return;
+            }
+
+            _state = targetState;
+            if (_appearanceComponent != null)
+            {
+                _appearanceComponent.SetData(PortalVisuals.State, _state);
+            }
+        }
+
+        private void releaseCooldown(IEntity entity)
+        {
+            if (immuneEntities.Contains(entity))
+            {
+                immuneEntities.Remove(entity);
+            }
+
+            if (_connectingTeleporter != null &&
+                _connectingTeleporter.TryGetComponent<ServerPortalComponent>(out var otherPortal))
+            {
+                otherPortal.immuneEntities.Remove(entity);
+            }
+        }
+
+        private IEnumerable<IEntity> GetPortableEntities()
+        {
+            // TODO: Performance optimisation?
```

Comment:
As a performance optimization, you can cut out the list creation and return and replace `portableEntities.Add(entity);` with `yield return entity;`. No need to allocate a list then.